### PR TITLE
Implement Limelight

### LINKED ISSUE #25: Limelight
We need to create the Limelight subsystem and at least one associated command.

We probably want to somewhat fuse [our approach from 2022](https://github.com/frc937/robot2022) with [our approach from 2023](https://github.com/frc937/robot2023/tree/cowtown2023). We should definitely keep the abstracted Limelight subsystem that makes it easy to add more Limelights to the bot, and there's no reason to get rid of the botpose-grabbing code, but we should start by using the trig-based method that just uses TX and TY to aim ourselves towards the amp.

Therefore, we should create a Limelight subsystem that can get all of the normal NetworkTables values (TX, TY, BotposeWPIBlue, BotposeWPIRed, etc.) and is abstracted such that creating multiple limelights is easy. We should probably feed the botpose into odometry somehow with swerveDriveInstance.addVisionMeasurement(Pose2d), but that can wait till later if we want.

We should also create an AimWithLimelight command that is abstracted. It should take the distance we want to be from the target, the height of the target off of the ground, the angle we want to be at relative to the target, the pipeline number, and maybe some identifier for which limelight to use as parameters. For the limelight mount height and angle, just use constants—but maybe add some way that we can only have one AimWithLimelight command that is still compatible with multiple Limelights. Maybe HashMaps in Constants that map the limelight names to their respective mount heights and angles?

Overall, just try to make this as flexible and as abstracted as possible. Command chaingun is not the way to go.

### PR DESCRIPTION
**Pull request by MarissaKoglesby**
closes #25

### CODE REVIEW

File: src/main/java/frc/robot/commands/AimWithLimelight.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      distanceFromTarget,
+      mountHeight,
+      mountAngle,
+      driveStrength,
+      speedLimit,
+      turnDoneThreshold,
+      distanceDoneThreshold,
+      upperHubTapeHeight;
+
+  /** Creates a new AimWithLimelight. */
```
[the remark this answers is not in the document]
Comment (willitcode) [reply src/main/java/frc/robot/commands/AimWithLimelight.java:36]: Also make sure there's more than just `@param`s but the `@param`s are the most important part

**Review by willitcode — CHANGES_REQUESTED** [review]
have fun

File: src/main/java/frc/robot/Constants.java [hunk 2]
```diff
@@ -74,4 +74,49 @@ public static class Intake {
...
+    /** Constants for aiming limelight. */
+    public static class AimingLimelight {
+      /** The number of degrees the limelight is mounted back from perfectly vertical */
+      public static final double MOUNT_ANGLE = 0;
+
+      /** The number of inches from the center of the Limelight lens to the floor */
+      public static final double MOUNT_HEIGHT = 0;
+
+      /** The number of inches from the retroreflective tape on the upper hub to the floor */
+      /**
```
Comment (willitcode) [inline src/main/java/frc/robot/Constants.java:89]: Fix this JavaDoc to actually describe what the var is. Also you can't have multiple JavaDoc comments on one field I don't think

File: src/main/java/frc/robot/RobotContainer.java [hunk 3]
```diff
@@ -51,6 +53,7 @@ public class RobotContainer {
   /** Singleton instance of {@link Intake} for the whole robot. */
   public static Intake intake = new Intake();
 
+  public static Limelight limelight = new Limelight("limelight");
```
Comment (willitcode) [inline src/main/java/frc/robot/RobotContainer.java:56]: JavaDoc

File: src/main/java/frc/robot/commands/AimWithLimelight.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+      distanceFromTarget,
+      mountHeight,
+      mountAngle,
+      driveStrength,
+      speedLimit,
+      turnDoneThreshold,
+      distanceDoneThreshold,
+      targetHeight;
+
+  /**
```
Comment (willitcode) [inline src/main/java/frc/robot/commands/AimWithLimelight.java:36]: Finish the JavaDoc

File: src/main/java/frc/robot/commands/AimWithLimelight.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+   * @param limelight The limelight to aim with.
```
Comment (willitcode) [inline src/main/java/frc/robot/commands/AimWithLimelight.java:37]: The Limelight subsystem to aim with

File: src/main/java/frc/robot/commands/AimWithLimelight.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+   * @param steerStrength How hard to turn towards the target; between 0 and 1.
+   * @param distanceFromTarget How far from in inches we want to be from the target when we
+   *     dispense.
+   * @param mountHeight The height of the limelight off the floor.
+   * @param mountAngle The number of degrees the limelight is mounted back from perfectly vertical.
```
Comment (willitcode) [inline src/main/java/frc/robot/commands/AimWithLimelight.java:42]: Does positive mean it's mounted pointing down from perfectly vertical or pointing up from perfectly vertical?

File: src/main/java/frc/robot/commands/AimWithLimelight.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+   * @param driveStrength How hard to drive towards the target.
```
Comment (willitcode) [inline src/main/java/frc/robot/commands/AimWithLimelight.java:43]: Clarify the range in this value

File: src/main/java/frc/robot/commands/AimWithLimelight.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+   * @param speedLimit Basic speed limit to make sure we don't drive too fast.
```
Comment (willitcode) [inline src/main/java/frc/robot/commands/AimWithLimelight.java:44]: Explain how to use this

File: src/main/java/frc/robot/commands/AimWithLimelight.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+   * @param turnDoneThreshold The threshold in angles when we consider the aiming done.
```
Comment (willitcode) [inline src/main/java/frc/robot/commands/AimWithLimelight.java:45]: Degrees or Radians?

File: src/main/java/frc/robot/commands/AimWithLimelight.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+      turnDoneThreshold,
+      distanceDoneThreshold,
+      targetHeight;
+
+  /**
+   * @param limelight The limelight to aim with.
+   * @param steerStrength How hard to turn towards the target; between 0 and 1.
+   * @param distanceFromTarget How far from in inches we want to be from the target when we
+   *     dispense.
+   * @param mountHeight The height of the limelight off the floor.
```
Comment (willitcode) [inline src/main/java/frc/robot/commands/AimWithLimelight.java:41]: Clarify that this is the distance to the center of the lens

File: src/main/java/frc/robot/subsystems/Limelight.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+  /* Note to future devs: we are NOT doing the whole setting values in NetworkTables to turn the Limelight
+   * LEDS on and off thing this year (turnOnLEDs() and turnOfflLEDs() in robot2022).
+   * For one, if we're not tracking reflective targets (which as of now we are not, we're just doing
+   * AprilTags), then we shouldn't need them on at all; for another, the correct way to turn the LEDs on
+   * and off is to make the default pipeline one with them off and then switch to a pipeline which has them
+   * on if we need to turn them on. This method prevents the whole "Limelight LEDs are stuck on until robot
+   * code starts!" thing.
+   */
+
+  /* This class may need logic to switch between pipelines based off of AprilTag pipeline restrictions. As
```
[the remark this answers is not in the document]
Comment (willitcode) [reply src/main/java/frc/robot/subsystems/Limelight.java:28]: Remove the comments

File: src/main/java/frc/robot/subsystems/Limelight.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+  private DoubleSubscriber pipelineSubcriber;
+  private DoublePublisher pipelinePublisher;
+  private String name;
+
+  private String fmtPath(String end) {
+    return "/" + name + "/" + end;
+  }
+
+  /**
+   * Creates a new Limelight. Should be run once from {@link frc.robot.RobotContainer}.
```
Comment (willitcode) [inline src/main/java/frc/robot/subsystems/Limelight.java:50]: Delete "Should be run once..."

File: src/main/java/frc/robot/subsystems/Limelight.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+   *
+   * @return A {@link Pose2d}; the robot's current position relative to the field in two dimensions.
+   */
+  public Pose2d getBotpose2d() {
+    double[] botpos = botposSubscriber.get();
+    return new Pose2d(botpos[0], botpos[1], Rotation2d.fromDegrees(botpos[5]));
+  }
+
+  /**
+   * Returns the current limelight pipeline.
```
Comment (willitcode) [inline src/main/java/frc/robot/subsystems/Limelight.java:143]: Limelight pipeline number

Also you need to go through and capitalize Limelight in all JavaDoc

**Review by willitcode — CHANGES_REQUESTED** [review]
Go through all JavaDoc and capitalize Limelight anywhere it isn't capitalized

File: src/main/java/frc/robot/RobotContainer.java [hunk 14]
```diff
@@ -51,6 +53,9 @@ public class RobotContainer {
   /** Singleton instance of {@link Intake} for the whole robot. */
   public static Intake intake = new Intake();
 
+  /** Singleton instance of {@link Limelight} for aiming. */
+  public static Limelight limelight = new Limelight("limelight");
```
Comment (willitcode) [inline src/main/java/frc/robot/RobotContainer.java:57]: Make this name a constant in the AimingLimelight subclass

File: src/main/java/frc/robot/commands/AimWithLimelight.java [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+      turnDoneThreshold,
+      distanceDoneThreshold,
+      targetHeight;
+
+  /**
+   * Creates a new command to aim with the Limelight.
+   *
+   * @param limelight The Limelight subsystem to aim with.
+   * @param steerStrength How hard to turn towards the target; between 0 and 1.
+   * @param distanceFromTarget How far from in inches we want to be from the target when we
```
Comment (willitcode) [inline src/main/java/frc/robot/commands/AimWithLimelight.java:41]: delete "when we dispense"—remember this class is abstracted

File: src/main/java/frc/robot/commands/AimWithLimelight.java [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+   *     dispense.
+   * @param mountHeight The height of the center of the Limelight lense off the floor.
```
Comment (willitcode) [inline src/main/java/frc/robot/commands/AimWithLimelight.java:43]: "lens"

File: src/main/java/frc/robot/commands/AimWithLimelight.java [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+   * @param mountAngle The number of degrees the Limelight is mounted back from perfectly vertical.
+   *     Positive means backwards.
```
Comment (willitcode) [inline src/main/java/frc/robot/commands/AimWithLimelight.java:45]: "Positive means rotated such that the lens is facing up, and not down"

**Review by willitcode — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/RobotContainer.java (hunk 14 above)
Comment (willitcode) [reply src/main/java/frc/robot/RobotContainer.java]: No seriously. Make the name of the Limelight a constant

File: src/main/java/frc/robot/commands/AimWithLimelight.java [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+   * Creates a new command to aim with the Limelight.
+   *
+   * @param limelight The Limelight subsystem to aim with.
+   * @param steerStrength How hard to turn towards the target; between 0 and 1.
+   * @param distanceFromTarget How far from in inches we want to be from the target.
+   * @param mountHeight The height of the center of the Limelight lens off the floor.
+   * @param mountAngle The number of degrees the Limelight is mounted back from perfectly vertical.
+   *     Positive means rotated such that the lens is facing up, and not down.
+   * @param driveStrength How hard to drive towards the target; between 0 and 1.
+   * @param speedLimit Basic speed limit to make sure we don't drive too fast. Percentage of max
```
Comment (willitcode) [inline src/main/java/frc/robot/commands/AimWithLimelight.java:46]: Clarify this is also between 0 and 1 so people don't pass it 70 and get confused why it's not working

**Review by willitcode — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Constants.java [hunk 19]
```diff
@@ -76,4 +76,47 @@ public static class Intake {
     /** Speed we want to run the Intake at. */
     public static final double INTAKE_MOTOR_SPEED = 0.5;
   }
+
+  /** Holds contstants for the Limelights. */
+  public static class Limelight {
+    /** Constants for aiming Limelight. */
+    public static class AimingLimelight {
```
Comment (Berdenson) [inline src/main/java/frc/robot/Constants.java:83]: Do we need an additional block for aiming the limelight? If so, I would advise moving LIMELIGHT_NAME down a block. If not, I'd remove it.

File: src/main/java/frc/robot/subsystems/Limelight.java [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+import edu.wpi.first.networktables.DoublePublisher;
+import edu.wpi.first.networktables.DoubleSubscriber;
+import edu.wpi.first.networktables.DoubleTopic;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+/** Subsystem for the Limelight 2+ that we use for vision. */
+public class Limelight extends SubsystemBase {
+
+  /* variable chaingun, I promise we use all of these */
```
Comment (Berdenson) [inline src/main/java/frc/robot/subsystems/Limelight.java:28]: Capitalization plz, maybe a period if you're feeling special. /lh

File: src/main/java/frc/robot/subsystems/Limelight.java [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+    this.pipelineSubcriber = pipelineTopic.subscribe(0.0);
+    this.pipelinePublisher = pipelineTopic.publish();
+    double[] defaultBotpos = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
+    botposSubscriber =
+        NetworkTableInstance.getDefault()
+            .getDoubleArrayTopic(fmtPath("botpose"))
+            .subscribe(defaultBotpos);
+  }
+
+  /* now its time for getter method chaingun, which I have to write manually because VS Code */
```
Comment (Berdenson) [inline src/main/java/frc/robot/subsystems/Limelight.java:64]: Capitalization. Also maybe remove the "write manually" "python script" parts if you want; keep if you value useless comedy.

File: src/main/java/frc/robot/subsystems/Limelight.java [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+   */
+  public Pose2d getBotpose2d() {
+    double[] botpos = botposSubscriber.get();
+    return new Pose2d(botpos[0], botpos[1], Rotation2d.fromDegrees(botpos[5]));
+  }
+
+  /**
+   * Returns the current Limelight pipeline number.
+   *
+   * @return the current Limelight pipeline number.
```
Comment (Berdenson) [inline src/main/java/frc/robot/subsystems/Limelight.java:138]: Capitalize.

File: src/main/java/frc/robot/subsystems/Limelight.java [hunk 23]
```diff
@@ -0,0 +1,159 @@
...
+  private String name;
+
+  private String fmtPath(String end) {
+    return "/" + name + "/" + end;
+  }
+
+  /**
+   * Creates a new Limelight.
+   *
+   * @param name The hostname of the Limelight
```
Comment (Berdenson) [inline src/main/java/frc/robot/subsystems/Limelight.java:45]: de-capatalize.

**Review by Berdenson — CHANGES_REQUESTED** [review]
Formatting and Grammar (:

**Review by willitcode — CHANGES_REQUESTED** [review]

**Review by willitcode — DISMISSED** [review]

**Review by willitcode — APPROVED** [review]

**Pull request merged**